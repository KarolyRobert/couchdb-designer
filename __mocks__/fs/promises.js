
'use strict';

const promises = jest.createMockFromModule('fs/promises');

promises.readFile = jest.fn();
promises.writeFile = jest.fn();
promises.readdir = jest.fn();
promises.access = jest.fn();
promises.stat = jest.fn();

module.exports = promises;
