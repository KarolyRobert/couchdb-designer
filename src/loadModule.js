import path from 'path';

export default function loadModule(directory,name){

    return new Promise((resolve,reject) => {
        let jsModule;
        try{
            jsModule = require(path.resolve(process.env.PWD,directory,name));
            resolve(jsModule);
        }catch(err){
            reject(err);
        }
    })
}