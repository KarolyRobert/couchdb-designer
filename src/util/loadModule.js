import path from 'path';

export default function loadModule(directory,name){

    return new Promise((resolve,reject) => {
        let jsModule = false;
        try{
            if(process.env.JEST_WORKER_ID){
                jest.useFakeTimers();
            }
            jsModule = require(path.resolve(process.env.PWD,directory,name));
         
            if(Object.keys(jsModule).length > 0){
                resolve(jsModule);
            }else{
                reject(new Error(`The module ${path.join(directory,name)} doesn't export anything! You must export function/s with module.exports = {...}`));
            }
        }catch(err){
           
            reject(err);
        }
    })
}