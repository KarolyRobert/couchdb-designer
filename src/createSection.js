import path from 'path';
import fs from 'fs/promises';
import createSectionFromFile from './createSectionFromFile';
import createSectionFromDirectory from './createSectionFromDirectory';

const createSection = (directory, name ) => {
    return new Promise((resolve, reject) => {
        let sectionPath = path.join(directory,name);
        fs.stat(sectionPath).then(stat => {
            if(stat.isFile() || stat.isDirectory()){
                if(stat.isFile()){
                    createSectionFromFile( directory, name ).then(result => resolve(result)
                    ,err => reject(err));
                }else{
                    createSectionFromDirectory( directory, name ).then(result => resolve(result)
                    ,err => reject(err));
                }
            }else{
                reject(`Bad structure! ${sectionPath} must be file or directory!`);
            }
        },err => reject(err));
    });
}

export default createSection;