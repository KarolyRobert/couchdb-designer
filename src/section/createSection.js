import path from 'path';
import fs from 'fs/promises';
import createDesignSectionFromFile from './createDesignSectionFromFile';
import createTestSectionFromFile from '../testing/createTestSectionFromFile';
import createSectionFromDirectory from './createSectionFromDirectory';


const createSection = (directory, name, contextProps) => {
   
    return new Promise((resolve, reject) => {
        let sectionPath = path.join(directory,name);
        fs.stat(sectionPath).then(fileStat => {
            if(fileStat.isFile() || fileStat.isDirectory()){
                if(fileStat.isFile()){
                    if(contextProps){
                        createTestSectionFromFile(directory, name, contextProps).then(resolve,reject);
                    }else{
                        createDesignSectionFromFile(directory, name).then(resolve,reject);
                    }
                }else{
                    createSectionFromDirectory( directory, name, contextProps).then(resolve,reject);
                }
            }else{
                reject(`Bad structure! ${sectionPath} must be file or directory!`);
            }
        },err => reject(err));
    });
    
}

export default createSection;