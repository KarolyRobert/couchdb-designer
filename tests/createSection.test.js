import createSection from '../src/section/createSection';
import fs from 'fs/promises';
import loadModule from '../src/util/loadModule';

jest.mock('../src/util/loadModule');
jest.mock('fs/promises');


 describe('createSection',() => {

    beforeEach(() => {
        jest.clearAllMocks();
    });


    test('success on file',() => {
        fs.stat.mockResolvedValue({isFile:() => true,isDirectory:() => false});
        loadModule.mockResolvedValue({
            validate_doc_update:function validate_doc_update(doc,req){}
        });
        return createSection('./root','validate_doc_update.js').then(result => {
            expect(result).toStrictEqual({validate_doc_update:'function (doc, req) {}'});
        });
    });


    test('success on directory',() => {
        fs.stat.mockResolvedValueOnce({isFile:() => false,isDirectory:() => true});
        fs.stat.mockResolvedValue({isFile:() => true,isDirectory:() => false});
        fs.readdir.mockResolvedValue(['egy.js','kettő.js']);
        loadModule.mockResolvedValueOnce({egy:function egy(){}});
        loadModule.mockResolvedValueOnce({kettő:function kettő(doc){ emit(doc._id,1)}});
        return createSection('design/root','shows').then(result => {
            expect(result).toMatchSnapshot();
        });
    });

    test('fail on not exist',() => {
        fs.stat.mockRejectedValue('no such file or directory');
        return createSection('./root','flie').catch(err => {
            expect(err).toBe('no such file or directory');
            expect(fs.readFile).not.toHaveBeenCalled();
        });
    });

    test('fail on not file or directory', ()=> {
        fs.stat.mockResolvedValue({isFile:() => false,isDirectory:() => false});
        return createSection('./root','flie').catch(err => {
            expect(err).toBe('Bad structure! root/flie must be file or directory!');
            expect(fs.readFile).not.toHaveBeenCalled();
        });
    });

});