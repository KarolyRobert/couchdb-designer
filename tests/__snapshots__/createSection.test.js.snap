// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSection success on directory 1`] = `
Object {
  "shows": Object {
    "egy": "function () {}",
    "kett≈ë": "function (doc) {
        emit(doc._id, 1);
      }",
  },
}
`;
